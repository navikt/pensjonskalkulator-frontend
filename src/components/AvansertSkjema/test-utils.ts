export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
