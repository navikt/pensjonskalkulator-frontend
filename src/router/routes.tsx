import { RouteObject, Navigate, Outlet } from 'react-router-dom'

import { PageFramework } from '@/components/common/PageFramework'
import { Beregning } from '@/pages/Beregning'
import { Forbehold } from '@/pages/Forbehold'
import { Henvisning } from '@/pages/Henvisning'
import { LandingPage } from '@/pages/LandingPage'
import { Personopplysninger } from '@/pages/Personopplysninger'
import { StepAFP } from '@/pages/StepAFP'
import { StepFeil } from '@/pages/StepFeil'
import { StepSamtykkeOffentligAFP } from '@/pages/StepSamtykkeOffentligAFP'
import { StepSamtykkePensjonsavtaler } from '@/pages/StepSamtykkePensjonsavtaler'
import { StepSivilstand } from '@/pages/StepSivilstand'
import { StepStart } from '@/pages/StepStart'
import { StepUfoeretrygdAFP } from '@/pages/StepUfoeretrygdAFP'
import { StepUtenlandsopphold } from '@/pages/StepUtenlandsopphold'
import { RouteErrorBoundary } from '@/router/RouteErrorBoundary'

import { paths } from './constants'
import {
  directAccessGuard,
  authenticationGuard,
  landingPageAccessGuard,
  stepStartAccessGuard,
  stepSivilstandAccessGuard,
  stepAFPAccessGuard,
  stepUfoeretrygdAFPAccessGuard,
  stepSamtykkeOffentligAFPAccessGuard,
} from './loaders'
import { IngenTilgang } from '@/pages/IngenTilgang'

export const routes: RouteObject[] = [
  {
    loader: authenticationGuard,
    element: (
      <PageFramework
        shouldShowLogo
        hasWhiteBg
        shouldRedirectNonAuthenticated={false}
      >
        <Outlet />
      </PageFramework>
    ),
    ErrorBoundary: RouteErrorBoundary,
    children: [
      {
        path: paths.root,
        element: <Navigate to={paths.login} replace />,
      },
      {
        loader: landingPageAccessGuard,
        path: paths.login,
        element: <LandingPage />,
      },
      {
        path: paths.personopplysninger,
        element: <Personopplysninger />,
      },
    ],
  },
  {
    loader: authenticationGuard,
    element: (
      <PageFramework>
        <Outlet />
      </PageFramework>
    ),
    ErrorBoundary: RouteErrorBoundary,
    children: [
      {
        loader: stepStartAccessGuard,
        path: paths.start,
        element: <StepStart />,
      },
      {
        path: `${paths.henvisning}/:id`,
        element: <Henvisning />,
      },
      {
        path: paths.forbehold,
        element: <Forbehold />,
      },
      {
        loader: stepSivilstandAccessGuard,
        path: paths.sivilstand,
        element: <StepSivilstand />,
      },
      {
        loader: directAccessGuard,
        path: paths.utenlandsopphold,
        element: <StepUtenlandsopphold />,
      },
      {
        loader: stepAFPAccessGuard,
        path: paths.afp,
        element: <StepAFP />,
      },
      {
        loader: stepUfoeretrygdAFPAccessGuard,
        path: paths.ufoeretrygdAFP,
        element: <StepUfoeretrygdAFP />,
      },
      {
        loader: stepSamtykkeOffentligAFPAccessGuard,
        path: paths.samtykkeOffentligAFP,
        element: <StepSamtykkeOffentligAFP />,
      },
      {
        loader: directAccessGuard,
        path: paths.samtykke,
        element: <StepSamtykkePensjonsavtaler />,
      },
      {
        loader: directAccessGuard,
        path: paths.uventetFeil,
        element: <StepFeil />,
      },
      {
        loader: directAccessGuard,
        path: paths.ingenTilgang,
        element: <IngenTilgang />,
      },
    ],
  },
  {
    loader: authenticationGuard,
    element: (
      <PageFramework isFullWidth hasWhiteBg>
        <Outlet />
      </PageFramework>
    ),
    ErrorBoundary: RouteErrorBoundary,
    children: [
      {
        loader: directAccessGuard,
        path: paths.beregningEnkel,
        element: <Beregning visning="enkel" />,
      },
      {
        loader: directAccessGuard,
        path: paths.beregningAvansert,
        element: <Beregning visning="avansert" />,
      },
    ],
  },
]
