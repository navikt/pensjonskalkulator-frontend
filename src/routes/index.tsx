export { routes, ROUTER_BASE_URL } from './routes'
