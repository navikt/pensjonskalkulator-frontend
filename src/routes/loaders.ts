import { LoaderFunction } from 'react-router-dom'

import { apiSlice } from '@/state/api/apiSlice'
import { store } from '@/state/store'

export const uttaksalderLoader = (async ({ request }) => {
  const promise = store.dispatch(
    apiSlice.endpoints.tidligsteUttaksalder.initiate()
  )
  request.signal.onabort = promise.abort

  return await promise
}) satisfies LoaderFunction
