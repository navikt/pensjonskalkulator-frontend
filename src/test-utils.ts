import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
