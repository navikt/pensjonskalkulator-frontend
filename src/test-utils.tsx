import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
