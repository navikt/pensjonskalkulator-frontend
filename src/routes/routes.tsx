import { redirect } from 'react-router'
import { RouteObject, Navigate, Outlet } from 'react-router-dom'

import { PageFramework } from '@/components/components/PageFramework'
import { Pensjonsberegning } from '@/containers/Pensjonsberegning'
import { LandingPage } from '@/routes/LandingPage'
import { RouteErrorBoundary } from '@/routes/RouteErrorBoundary'
import { Step1 } from '@/routes/stegvisning/Step1'
import { Step2 } from '@/routes/stegvisning/Step2'
import { Step3 } from '@/routes/stegvisning/Step3'
import { step3loader } from '@/routes/stegvisning/Step3/utils'
import { Step4 } from '@/routes/stegvisning/Step4'
import { Step5 } from '@/routes/stegvisning/Step5'
import { store } from '@/state/store'

export const BASE_PATH = '/pensjon/kalkulator'

export const externalsUrls = {
  dinPensjon: 'http://www.nav.no/pensjon',
}

export const paths = {
  root: '/',
  login: '/login',
  start: '/start',
  samtykke: '/samtykke',
  offentligTp: '/offentlig-tp',
  afp: '/afp',
  sivilstand: '/sivilstand',
  beregning: '/beregning',
} as const

const directAccessGuard = async () => {
  // Dersom ingen kall er registrert i store betyr det at brukeren prøver å aksessere en url direkte
  if (
    store.getState().api.queries === undefined ||
    Object.keys(store.getState().api.queries).length === 0
  ) {
    return redirect(paths.start)
  }
  return null
}

export const routes: RouteObject[] = [
  {
    element: (
      <PageFramework>
        <Outlet />
      </PageFramework>
    ),
    ErrorBoundary: RouteErrorBoundary,
    children: [
      {
        path: paths.root,
        element: <Navigate to={paths.start} replace />,
      },
      {
        path: paths.login,
        element: <LandingPage />,
      },
      {
        path: paths.start,
        element: <Step1 />,
      },
      {
        path: paths.samtykke,
        loader: directAccessGuard,
        element: <Step2 />,
      },
      {
        path: paths.offentligTp,
        loader: step3loader,
        element: <Step3 />,
      },
      {
        path: paths.afp,
        loader: directAccessGuard,
        element: <Step4 />,
      },
      {
        path: paths.sivilstand,
        loader: directAccessGuard,
        element: <Step5 />,
      },
    ],
  },
  {
    path: paths.beregning,
    loader: directAccessGuard,
    element: (
      <PageFramework isFullWidth>
        <Pensjonsberegning />
      </PageFramework>
    ),
    ErrorBoundary: RouteErrorBoundary,
  },
]
