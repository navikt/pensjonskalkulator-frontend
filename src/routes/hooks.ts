import { useLoaderData } from 'react-router-dom'

import { uttaksalderLoader } from '@/routes/loaders'

export const useTidligsteUttaksalder = () => {
  return useLoaderData() as Awaited<ReturnType<typeof uttaksalderLoader>>
}
