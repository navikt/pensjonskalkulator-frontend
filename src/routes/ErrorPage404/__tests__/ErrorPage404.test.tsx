import * as ReactRouterUtils from 'react-router'

import { describe, it, vi } from 'vitest'

import { ErrorPage404 } from '..'
import { externalUrls } from '@/routes'
import { render, screen, userEvent } from '@/test-utils'

const realLocation = window.location

describe('ErrorPage404', () => {
  afterEach(() => {
    window.location = realLocation
  })

  it('rendrer med riktig tekst og knapper', () => {
    const { asFragment } = render(<ErrorPage404 />)
    expect(screen.getAllByRole('button')).toHaveLength(2)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sender brukeren til forrige siden når brukeren klikker på første knapp', async () => {
    const user = userEvent.setup()
    const navigateMock = vi.fn()
    vi.spyOn(ReactRouterUtils, 'useNavigate').mockImplementation(
      () => navigateMock
    )
    render(<ErrorPage404 />)

    await user.click(screen.getByText('errorpage.404.button.primary'))
    expect(navigateMock).toHaveBeenCalledWith(-1)
  })

  it('sender brukeren til Din Pensjon når brukeren klikker på andre knapp', async () => {
    const user = userEvent.setup()
    global.window = Object.create(window)
    const url = 'http://dummy.com'
    Object.defineProperty(window, 'location', {
      value: {
        href: url,
      },
      writable: true,
    })

    render(<ErrorPage404 />)

    await user.click(screen.getByText('errorpage.404.button.secondary'))
    expect(window.location.href).toBe(externalUrls.dinPensjon)
  })
})
