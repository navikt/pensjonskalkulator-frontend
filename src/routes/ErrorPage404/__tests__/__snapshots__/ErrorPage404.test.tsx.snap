// Vitest Snapshot v1

exports[`ErrorPage404 > rendrer med riktig tekst og knapper 1`] = `
<DocumentFragment>
  <main
    class="_main_4532c9 _main__isFramed_4532c9"
  >
    <div
      class="_headerGroup_4532c9"
    >
      <h1
        class="_headerGroupTitle_4532c9 _headerGroupTitle__isFramed_4532c9 navds-heading navds-heading--large"
      >
        application.title
      </h1>
    </div>
    <section
      class="_section_645946 _section__largePadding_645946"
      data-testid="error-page-404"
    >
      <h2
        class="navds-heading navds-heading--medium navds-typo--spacing"
      >
        errorpage.404.title
      </h2>
      <div
        class="navds-list"
      >
        <ul>
          <li
            class="navds-list__item"
          >
            <div
              class="navds-list__item-marker--bullet"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="6"
                role="img"
                viewBox="0 0 6 6"
                width="6"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="currentColor"
                  height="6"
                  rx="3"
                  width="6"
                />
              </svg>
            </div>
            <div
              class="navds-list__item-content navds-body-short navds-body-short--small"
            >
              <p
                class="navds-body-short"
              >
                errorpage.404.list_item1
              </p>
            </div>
          </li>
          <li
            class="navds-list__item"
          >
            <div
              class="navds-list__item-marker--bullet"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="6"
                role="img"
                viewBox="0 0 6 6"
                width="6"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="currentColor"
                  height="6"
                  rx="3"
                  width="6"
                />
              </svg>
            </div>
            <div
              class="navds-list__item-content navds-body-short navds-body-short--small"
            >
              <p
                class="navds-body-short"
              >
                errorpage.404.list_item2
              </p>
            </div>
          </li>
        </ul>
      </div>
      <button
        class="_button_bc1ea0 navds-button navds-button--primary navds-button--medium"
      >
        <span
          class="navds-label"
        >
          errorpage.404.button.primary
        </span>
      </button>
      <button
        class="_button_bc1ea0 navds-button navds-button--secondary navds-button--medium"
      >
        <span
          class="navds-label"
        >
          errorpage.404.button.secondary
        </span>
      </button>
    </section>
  </main>
</DocumentFragment>
`;
